### PR TITLE
General Updates

### PR DESCRIPTION
**Pull request by brianreavis**
- Official support for Node 4/5
- Better error handling (there were a few places where errors weren't being handled, which caused gulp to silently stop without any info)
- Updated [spritesmith](https://github.com/Ensighten/spritesmith) `0.18.0` -> `3.1.0`
- Updated [lodash](https://www.npmjs.com/package/lodash) `2.4.1` -> `4.1.1`
- Updated [through2](https://www.npmjs.com/package/through2) `0.4.1` -> `2.0.1`
- Updated [async](https://www.npmjs.com/package/async) `0.2.10` -> `2.0.0-rc.3`
- Updated [vinyl](https://github.com/gulpjs/vinyl) `0.2.3` -> `1.1.1`
- Dropped [Q](https://github.com/kriskowal/q) in favor of [es6-promise](https://www.npmjs.com/package/es6-promise) (easier to drop later when promises are mainstream)

p.s. sorry in advance for style changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,14 +7,20 @@
...
 
-Install with [npm](https://npmjs.org/package/gulp-sprite-generator)
+Install with [npm](https://npmjs.org/package/gulp-sprite-generator):
 
+```sh
+$ npm install --save-dev gulp-sprite-generator
 ```
-npm install --save-dev gulp-sprite-generator
+
+You also should install a spritesmith engine:
```
Comment (gobwas) [inline README.md:20]: Why user should install it?
Spritesmith will install it via it's package.json, [isnt it](https://github.com/Ensighten/spritesmith/blob/master/package.json#L35)?

File: index.js [hunk 2]
```diff
@@ -113,65 +105,45 @@ var getImages = (function() {
...
-                                return reject(err);
-                            }
-
-                            resolve(images);
-                        }
-                    );
+                return new Promise(function(resolve, reject) {
+                    async.filter(images, function(image, callback) {
+                        async.reduce(options.filter, true, function(status, filter, callback) {
+                            if (!status) return callback(null, false);
```
Comment (gobwas) [inline index.js:118]: Could you wrap `return callback...` in a curly brackets? It will be more readable, I think.

File: index.js [hunk 3]
```diff
@@ -113,65 +105,45 @@ var getImages = (function() {
...
+                            Promise.resolve(filter(image))
+                                .then(function(status) { callback(null, status); })
```
Comment (gobwas) [inline index.js:120]: It is a not good coding style to have one lined functions.

File: index.js [hunk 4]
```diff
@@ -113,65 +105,45 @@ var getImages = (function() {
...
+                                .catch(callback);
+                        }, callback);
+                    }, function(err, filteredImages) {
+                        if (err) return reject(err);
```
Comment (gobwas) [inline index.js:124]: The same as above.

File: index.js [hunk 5]
```diff
@@ -113,65 +105,45 @@ var getImages = (function() {
...
                         function(images, groupBy, next) {
                             async.map(images, function(image, done) {
-                                Q(groupBy(image))
+                                Promise.resolve(groupBy(image))
                                     .then(function(group) {
-                                        if (group) {
-                                            image.group.push(group);
-                                        }
-
+                                        if (group) image.group.push(group);
```
Comment (gobwas) [inline index.js:139]: No.

File: index.js [hunk 6]
```diff
@@ -113,65 +105,45 @@ var getImages = (function() {
...
                                     })
                                     .catch(done);
                             }, next);
                         },
                         function(err, images) {
-                            if (err) {
-                                return reject(err);
-                            }
-
+                            if (err) return reject(err);
```
Comment (gobwas) [inline index.js:146]: Again, no.

File: index.js [hunk 7]
```diff
@@ -181,65 +153,60 @@ var getImages = (function() {
 })();
 
 var callSpriteSmithWith = (function() {
-    var GROUP_DELIMITER = ".",
-        GROUP_MASK = "*";
+    var GROUP_DELIMITER = ".", GROUP_MASK = "*";
```
Comment (gobwas) [inline index.js:156]: If you split the `var`, use `var` at each line.

File: index.js [hunk 8]
```diff
@@ -181,65 +153,60 @@ var getImages = (function() {
...
+                    src: images.map(function(image) {
+                        return image.path;
+                    })
                 });
 
                 // enlarge padding, if its retina
-                if (_.every(images, function(image) {return image.isRetina})) {
-                    ratio = _.chain(images).flatten('retinaRatio').unique().value();
-                    if (ratio.length == 1) {
+                if (_.every(images, function(image) { return image.isRetina; })) {
```
Comment (gobwas) [inline index.js:182]: Could you refactor this to avoid one line functions?

File: index.js [hunk 9]
```diff
@@ -181,65 +153,60 @@ var getImages = (function() {
...
+                    var ratio = _.chain(images).flatten('retinaRatio').uniq().value();
+                    if (ratio.length === 1) {
```
Comment (gobwas) [inline index.js:184]: Does `[].length` could be a non-numeric? )

File: index.js [hunk 10]
```diff
@@ -181,65 +153,60 @@ var getImages = (function() {
...
-                return Q.nfcall(spritesmith, config).then(function(result) {
-                    tmp = tmp.split(GROUP_DELIMITER);
-                    tmp.shift();
-
-                    // append info about sprite group
-                    result.group = tmp.map(mask(false));
-
-                    return result;
+                // validate config
+                try { new Spritesmith(config); }
```
Comment (gobwas) [inline index.js:190]:
``` js
try {
...
} catch (err) {
...
}
```

File: index.js [hunk 11]
```diff
@@ -181,65 +153,60 @@ var getImages = (function() {
...
+                catch (e) { return Promise.reject(e);}
+
+                return new Promise(function(resolve, reject) {
+                    Spritesmith.run(config, function(err, result) {
+                        if (err) return reject(err);
```
Comment (gobwas) [inline index.js:195]: Not okay.

**Comment by brianreavis** [comment]
Going to stop you there – this isn't going to be worth either of our time.

File: test/test.js [hunk 12]
```diff
@@ -450,31 +456,22 @@ describe('gulp-sprite-generator', function(){
...
+            console.error('Encountered stream error:');
+            console.error(e.stack);
+            errors.push(err);
+        });
         stream.on('finish', function() {
             try {
-                assert.equal(1, piped.img,  "No piped data in img stream");
-                assert.equal(1, piped.css,  "No piped data in css stream");
-                assert.equal(1, piped.main, "No piped data in main stream");
+                assert.equal(1, piped.img, 'No piped data in img stream');
```
Comment (gobwas) [inline test/test.js:472]: Did not agree with it.

**Comment by gobwas** [comment]
Thanks for your work and PR, but I did not understand the sense of changing the coding style without meaningful reason. If you want, you could refactor the logic as you did with filtering images and it's ok and welcome. Also Im agree with the one lined `var`. But another code style changes did not makes sense for me.

**Pull request closed without merging**